pkg/pillar/cmd/downloader/downloader.go

```diff
@@ -113,151 +88,41 @@ func Run() {
-	subDatastoreConfig.Activate()
-
-	pubGlobalDownloadStatus, err := pubsub.Publish(agentName,
-		types.GlobalDownloadStatus{})
-	if err != nil {
-		log.Fatal(err)
-	}
-	ctx.pubGlobalDownloadStatus = pubGlobalDownloadStatus
-
-	// Set up our publications before the subscriptions so ctx is set
-	pubAppImgStatus, err := pubsub.PublishScope(agentName, types.AppImgObj,
-		types.DownloaderStatus{})
-	if err != nil {
-		log.Fatal(err)
-	}
-	ctx.pubAppImgStatus = pubAppImgStatus
-	pubAppImgStatus.ClearRestarted()
-
-	pubBaseOsStatus, err := pubsub.PublishScope(agentName, types.BaseOsObj,
-		types.DownloaderStatus{})
+	// set up any state needed by handler functions
+	err = ctx.registerHandlers()
 	if err != nil {
 		log.Fatal(err)
 	}
-	ctx.pubBaseOsStatus = pubBaseOsStatus
-	pubBaseOsStatus.ClearRestarted()
-
-	pubCertObjStatus, err := pubsub.PublishScope(agentName, types.CertObj,
-		types.DownloaderStatus{})
-	if err != nil {
-		log.Fatal(err)
-	}
-	ctx.pubCertObjStatus = pubCertObjStatus
-	pubCertObjStatus.ClearRestarted()
-
-	subAppImgConfig, err := pubsub.SubscribeScope("zedmanager",
-		types.AppImgObj, types.DownloaderConfig{}, false, &ctx)
-	if err != nil {
-		log.Fatal(err)
-	}
-	subAppImgConfig.ModifyHandler = handleAppImgModify
-	subAppImgConfig.CreateHandler = handleAppImgCreate
-	subAppImgConfig.DeleteHandler = handleAppImgDelete
-	ctx.subAppImgConfig = subAppImgConfig
-	subAppImgConfig.Activate()
-
-	subBaseOsConfig, err := pubsub.SubscribeScope("baseosmgr",
-		types.BaseOsObj, types.DownloaderConfig{}, false, &ctx)
-	if err != nil {
-		log.Fatal(err)
-	}
-	subBaseOsConfig.ModifyHandler = handleBaseOsModify
-	subBaseOsConfig.CreateHandler = handleBaseOsCreate
-	subBaseOsConfig.DeleteHandler = handleBaseOsDelete
-	ctx.subBaseOsConfig = subBaseOsConfig
-	subBaseOsConfig.Activate()
-
-	subCertObjConfig, err := pubsub.SubscribeScope("baseosmgr",
-		types.CertObj, types.DownloaderConfig{}, false, &ctx)
-	if err != nil {
-		log.Fatal(err)
-	}
-	subCertObjConfig.ModifyHandler = handleCertObjModify
-	subCertObjConfig.CreateHandler = handleCertObjCreate
-	subCertObjConfig.DeleteHandler = handleCertObjDelete
-	ctx.subCertObjConfig = subCertObjConfig
-	subCertObjConfig.Activate()
-
-	pubAppImgStatus.SignalRestarted()
-	pubBaseOsStatus.SignalRestarted()
-	pubCertObjStatus.SignalRestarted()
 
 	// First wait to have some management ports with addresses
 	// Looking at any management ports since we can do baseOS download over all
 	// Also ensure GlobalDownloadConfig has been read
 	for types.CountLocalAddrAnyNoLinkLocal(ctx.deviceNetworkStatus) == 0 ||
 		ctx.globalConfig.MaxSpace == 0 {
 		log.Infof("Waiting for management port addresses or Global Config\n")
+		start := agentlog.StartTime()
+		checkMax := true
 
 		select {
-		case change := <-subGlobalConfig.C:
-			start := agentlog.StartTime()
-			subGlobalConfig.ProcessChange(change)
-			agentlog.CheckMaxTime(agentName, start)
+		case change := <-ctx.subGlobalConfig.C:
```

Comment:
Wondering of we can make it a generic routine that takes a slice of Subscriptions to process.. This pattern is repeated a lot in lots of agents.. Something like..

SelectAndProcessSubscriptions([Sub1, Sub2, Sub3 etc..])

All these Subscriptions have the ProcessChange handler anyways..
Maybe even make this a utility function in Agent.go ??

Reply:
Yes, I very much want to. I was going to move it to `downloaderContext` as an associated `func`. I stopped myself for the same reason as above. It was easier to see and reason about these changes - which just involve extracting repeated code - and then simplify the whole channel thing in one or more subsequent PRs.

Reply:
Agree.. Subsequent PRs is fine..